File: cli/tbonlineinx.c
/****************************************************************************
** (c) Copyright IBM Corp. 2007 All rights reserved.
** 
** The following sample of source code ("Sample") is owned by International 
** Business Machines Corporation or one of its subsidiaries ("IBM") and is 
** copyrighted and licensed, not sold. You may use, copy, modify, and 
** distribute the Sample in any form without payment to IBM, for the purpose of 
** assisting you in the development of your applications.
** 
** The Sample code is provided to you on an "AS IS" basis, without warranty of 
** any kind. IBM HEREBY EXPRESSLY DISCLAIMS ALL WARRANTIES, EITHER EXPRESS OR 
** IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF 
** MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. Some jurisdictions do 
** not allow for the exclusion or limitation of implied warranties, so the above 
** limitations or exclusions may not apply to you. IBM shall not be liable for 
** any damages you suffer as a result of using, copying, modifying or 
** distributing the Sample, even if IBM has been advised of the possibility of 
** such damages.
*****************************************************************************
**
** SOURCE FILE NAME: tbonlineinx.c                                      
**                                                                        
** SAMPLE:  How to create and reorg indexes on a table
**
** CLI FUNCTIONS USED:
**         SQLAllocHandle -- Allocate Handle
**         SQLExecDirect -- Execute a Statement Directly
**         SQLFetch -- Fetch Next Row
**         SQLFreeHandle -- Free Handle Resources
**         SQLGetData -- Get Data From a Column                
**         SQLSetConnectAttr -- Set Connection Attributes
**
** DB2 APIs USED:
**         db2Reorg -- Reorganize a Table or Index
**
** SQL STRUCTURES USED:
**         sqlca
**         db2ReorgStruct 
**
** OUTPUT FILE: tbonlineinx.out (available in the online documentation)
*****************************************************************************
**
** For more information on the sample programs, see the README file.
**
** For information on developing CLI applications, see the CLI Guide
** and Reference.
**
** For information on using SQL statements, see the SQL Reference.
**
** For the latest information on programming, building, and running DB2 
** applications, visit the DB2 application development website: 
**     http://www.software.ibm.com/data/db2/udb/ad
****************************************************************************/

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <sqlcli1.h>
#include <db2ApiDf.h>
#include "utilcli.h" /* Header file for CLI sample code */

int CreateIndex(SQLHANDLE);
int ReorgIndex(SQLHANDLE);
int DropIndex(SQLHANDLE);
int SchemaNameGet(SQLHANDLE);

char schemaName[10];
char tableName[10];

int main(int argc, char *argv[])
{
  SQLRETURN cliRC = SQL_SUCCESS;
  int rc = 0;
  SQLHANDLE henv; /* environment handle */
  SQLHANDLE hdbc; /* connection handles */

  char dbAlias[SQL_MAX_DSN_LENGTH + 1];
  char user[MAX_UID_LENGTH + 1];
  char pswd[MAX_PWD_LENGTH + 1];

  /* checks the command line arguments */
  rc = CmdLineArgsCheck1(argc, argv, dbAlias, user, pswd);
  if (rc != 0)
  {
    return rc;
  }

  printf("\nTHIS SAMPLE SHOWS HOW TO CREATE AND REORG ONLINE INDEXES\n");
  printf("ON TABLES.\n");
  
  /* initialize the CLI application by calling a helper
     utility function defined in utilcli.c */
  rc = CLIAppInit(dbAlias,
                  user,
                  pswd,
                  &henv,
                  &hdbc,
                  (SQLPOINTER)SQL_AUTOCOMMIT_ON);
  if (rc != 0)
  {
    return rc;
  }
  
  /* create online index on a table */
  rc = CreateIndex(hdbc);
  if (rc != 0)
  {
    return rc;
  }  

  /* reorg online index on a table */
  rc = ReorgIndex(hdbc);
  if (rc != 0)
  {
    return rc;
  }
  
  /* terminate the CLI application by calling a helper
     utility function defined in utilcli.c */
  rc = CLIAppTerm(&henv, &hdbc, dbAlias);

  return rc;
} /* main */

/* to create an index on a table with different levels
   of access to the table like read-write, read-only, no access */
int CreateIndex(SQLHANDLE hdbc)
{
  SQLRETURN cliRC = SQL_SUCCESS;
  int rc = 0;
  SQLHANDLE hstmt; /* statement handle */
  
  SQLCHAR *stmt_inx = (SQLCHAR *)"CREATE INDEX INDEX1 ON "
                                 "EMPLOYEE (LASTNAME ASC)";
   
  SQLCHAR *stmt_lsm = (SQLCHAR *)"LOCK TABLE EMPLOYEE IN SHARE MODE";

  SQLCHAR *stmt_lem = (SQLCHAR *)"LOCK TABLE EMPLOYEE IN EXCLUSIVE MODE";

  printf("\n-----------------------------------------------------------");
  printf("\nUSE THE CLI FUNCTIONS\n");
  printf("  SQLSetConnectAttr\n");
  printf("  SQLAllocHandle\n");
  printf("  SQLExecDirect\n");
  printf("  SQLFreeHandle\n");
  printf("TO CREATE INDEX WITH DIFFERENT LEVELS OF ACCESS:\n");
  
  /* set AUTOCOMMIT on */
  cliRC = SQLSetConnectAttr(hdbc,
                            SQL_ATTR_AUTOCOMMIT,
                            (SQLPOINTER)SQL_AUTOCOMMIT_ON,
                            SQL_NTS);
  DBC_HANDLE_CHECK(hdbc, cliRC);

  /* create an online index with read-write access to the table */
  printf("\nTo create an index on a table allowing read-write access\n");
  printf("to the table, use the following SQL command:\n\n");
  printf("  Directly execute the statement\n");
  printf("    CREATE INDEX INDEX1 ON EMPLOYEE (LASTNAME ASC)\n");
  
  /* allocate a statement handle */
  cliRC = SQLAllocHandle(SQL_HANDLE_STMT, hdbc, &hstmt);
  DBC_HANDLE_CHECK(hdbc, cliRC);

  /* create index index1 */
  cliRC = SQLExecDirect(hstmt, stmt_inx, SQL_NTS);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  /* free the statement handle */
  cliRC = SQLFreeHandle(SQL_HANDLE_STMT, hstmt);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  rc = DropIndex(hdbc);
  if (rc != 0)
  {
    return rc;
  }

  /* create an index on a table while allowing only read access to it */
  printf("\nTo create an index on a table allowing only read access\n");
  printf("to the table, use the following two SQL commands:\n\n"); 
  printf("  Directly execute the statements\n");
  printf("    LOCK TABLE EMPLOYEE IN SHARE MODE\n");
  printf("    CREATE INDEX INDEX1 ON EMPLOYEE (LASTNAME ASC)\n");

  /* allocate a statement handle */
  cliRC = SQLAllocHandle(SQL_HANDLE_STMT, hdbc, &hstmt);
  DBC_HANDLE_CHECK(hdbc, cliRC);

  /* lock table in share mode */
  cliRC = SQLExecDirect(hstmt, stmt_lsm, SQL_NTS);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  /* free the statement handle */
  cliRC = SQLFreeHandle(SQL_HANDLE_STMT, hstmt);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);
  
  /* allocate a statement handle */
  cliRC = SQLAllocHandle(SQL_HANDLE_STMT, hdbc, &hstmt);
  DBC_HANDLE_CHECK(hdbc, cliRC);

  /* create index index1 */
  cliRC = SQLExecDirect(hstmt, stmt_inx, SQL_NTS);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  /* free the statement handle */
  cliRC = SQLFreeHandle(SQL_HANDLE_STMT, hstmt);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);
 
  rc = DropIndex(hdbc);
  if (rc != 0)
  {
    return rc;
  }
 
  /* create an online index allowing no access to the table */
  printf("\nTo create an index on a table allowing no access to the \n");
  printf("table (only uncommitted readers allowed), use the \n");
  printf("following two SQL statements:\n\n");
  printf("  Directly execute the statements\n");
  printf("    LOCK TABLE EMPLOYEE IN EXCLUSIVE MODE\n");
  printf("    CREATE INDEX INDEX1 ON EMPLOYEE (LASTNAME ASC)\n");
  
  /* allocate a statement handle */
  cliRC = SQLAllocHandle(SQL_HANDLE_STMT, hdbc, &hstmt);
  DBC_HANDLE_CHECK(hdbc, cliRC);

  /* lock table in exclusive mode */
  cliRC = SQLExecDirect(hstmt, stmt_lem, SQL_NTS);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  /* free the statement handle */
  cliRC = SQLFreeHandle(SQL_HANDLE_STMT, hstmt);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  /* allocate a statement handle */
  cliRC = SQLAllocHandle(SQL_HANDLE_STMT, hdbc, &hstmt);
  DBC_HANDLE_CHECK(hdbc, cliRC);

  /* create index index1 */
  cliRC = SQLExecDirect(hstmt, stmt_inx, SQL_NTS);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  /* free the statement handle */
  cliRC = SQLFreeHandle(SQL_HANDLE_STMT, hstmt);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  return rc;
} /* CreateIndex */

/* to reorg an index on a table with different levels of 
   access to the table like read-write, read-only, no access */
int ReorgIndex(SQLHANDLE hdbc)
{
  int rc = 0;
  struct sqlca sqlca;
  char fullTableName[258];
  db2ReorgStruct paramStruct;
  db2Uint32 versionNumber = db2Version970;

  printf("\n-----------------------------------------------------------");

  printf("\nUSE THE DB2 API:\n");
  printf("  db2Reorg -- Reorganize a Table or Index\n");
  printf("TO REORGANIZE A TABLE OR INDEX.\n");  

  /* get fully qualified name of the table */
  strcpy(tableName, "EMPLOYEE");

  rc = SchemaNameGet(hdbc);
  if (rc != 0)
  {
    return rc;
  }

  strcpy(fullTableName, schemaName);
  strcat(fullTableName, ".");
  strcat(fullTableName, tableName);

  printf("  Reorganize all indexes defined on table : %s\n", fullTableName);

  /* setup parameters */
  memset(&paramStruct, '\0', sizeof(paramStruct));
  paramStruct.reorgObject.tableStruct.pTableName = fullTableName;
  paramStruct.reorgObject.tableStruct.pOrderByIndex = NULL;
  paramStruct.reorgObject.tableStruct.pSysTempSpace = NULL;
  paramStruct.reorgType = DB2REORG_OBJ_INDEXESALL;
  paramStruct.nodeListFlag = DB2_ALL_NODES;
  paramStruct.numNodes = 0;
  paramStruct.pNodeList = NULL;

  printf("  \nReorganize the indexes on a table allowing read-write\n");
  printf("  access to the table (set reorgFlags to DB2REORG_ALLOW_WRITE)\n");

  paramStruct.reorgFlags = DB2REORG_ALLOW_WRITE;

  /* reorganize index */
  rc = db2Reorg(versionNumber, &paramStruct, &sqlca);
  DB2_API_CHECK("index -- reorganize");

  printf("  \nReorganize the indexes on a table allowing read-only\n");
  printf("  access to the table (set reorgFlags to DB2REORG_ALLOW_READ)\n");

  paramStruct.reorgFlags = DB2REORG_ALLOW_READ;

  /* reorganize index */
  rc = db2Reorg(versionNumber, &paramStruct, &sqlca);
  DB2_API_CHECK("index -- reorganize");

  printf("  \nReorganize the indexes on a table allowing no access\n");
  printf("  to the table (set reorgFlags to DB2REORG_ALLOW_NONE)\n");

  paramStruct.reorgFlags = DB2REORG_ALLOW_NONE;

  /* reorganize index */
  rc = db2Reorg(versionNumber, &paramStruct, &sqlca);
  DB2_API_CHECK("index -- reorganize");

  rc = DropIndex(hdbc);
  if (rc != 0)
  {
    return rc;
  }

  return rc;
} /* ReorgIndex */

/* to drop the index on a table */
int DropIndex(SQLHANDLE hdbc)
{
  SQLRETURN cliRC = SQL_SUCCESS;
  int rc = 0;
  SQLHANDLE hstmt; /* statement handle */
  
  SQLCHAR *stmt = (SQLCHAR *) " DROP INDEX index1 ";

  printf("\nUSE THE CLI FUNCTIONS\n");
  printf("  SQLAllocHandle\n");
  printf("  SQLExecDirect\n");
  printf("  SQLFreeHandle\n");
  printf("  SQLSetConnectAttr\n");
  printf("TO DROP THE INDEX:\n");
  
  /* drop the indexes */
  printf("\n  Directly execute the statement\n");
  printf("    DROP INDEX index1\n");

  /* allocate a statement handle */
  cliRC = SQLAllocHandle(SQL_HANDLE_STMT, hdbc, &hstmt);
  DBC_HANDLE_CHECK(hdbc, cliRC);

  /* drop the index */
  cliRC = SQLExecDirect(hstmt, stmt, SQL_NTS);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  /* free the statement handle */
  cliRC = SQLFreeHandle(SQL_HANDLE_STMT, hstmt);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC); 
  
  printf("\n-----------------------------------------------------------");

  return rc;
} /* DropIndex */

/* gets the schema name of the table */
int SchemaNameGet(SQLHANDLE hdbc)
{
  SQLRETURN cliRC = SQL_SUCCESS;
  int rc = 0;
  SQLHANDLE hstmt; /* statement handle */

  SQLCHAR *stmt = (SQLCHAR *)"SELECT tabschema FROM syscat.tables "
                             "WHERE tabname = 'EMPLOYEE' ";

  struct
  {
    SQLINTEGER ind;
    SQLCHAR val[15];
  }
  tabschema; /* variable to be bound to the tabschema column */

  printf("\nUSE THE CLI FUNCTIONS\n");
  printf("  SQLSetConnectAttr\n");
  printf("  SQLAllocHandle\n");
  printf("  SQLExecDirect\n");
  printf("  SQLFetch\n");
  printf("  SQLFreeHandle\n");
  printf("  SQLGetData\n");
  printf("TO GET SCHEMA NAME OF THE TABLE\n\n");

  /* set AUTOCOMMIT on */
  cliRC = SQLSetConnectAttr(hdbc,
                            SQL_ATTR_AUTOCOMMIT,
                            (SQLPOINTER)SQL_AUTOCOMMIT_ON,
                            SQL_NTS);
  DBC_HANDLE_CHECK(hdbc, cliRC);

  /* allocate a statement handle */
  cliRC = SQLAllocHandle(SQL_HANDLE_STMT, hdbc, &hstmt);
  DBC_HANDLE_CHECK(hdbc, cliRC);

  /* directly execute the statement */
  cliRC = SQLExecDirect(hstmt, stmt, SQL_NTS);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  /* fetch each row and display */
  cliRC = SQLFetch(hstmt);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  if (cliRC == SQL_NO_DATA_FOUND)
  {
    printf("\n  Data not found.\n");
  }
  while (cliRC != SQL_NO_DATA_FOUND)
  {
    /* get data from column */
    cliRC = SQLGetData(hstmt,
                       1,
                       SQL_C_CHAR,
                       tabschema.val,
                       10,
                       &tabschema.ind);
    STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);
  }
  
  strcpy(schemaName, (char *)tabschema.val);
   
  /* get rid of spaces from the end of schemaName */
  strtok(schemaName, " ");
  
  /* free the statement handle */
  cliRC = SQLFreeHandle(SQL_HANDLE_STMT, hstmt);
  STMT_HANDLE_CHECK(hstmt, hdbc, cliRC);

  return rc;
} /* SchemaNameGet */

File: java/jdbc/TbOnlineInx.java
//***************************************************************************
// (c) Copyright IBM Corp. 2007 All rights reserved.
// 
// The following sample of source code ("Sample") is owned by International 
// Business Machines Corporation or one of its subsidiaries ("IBM") and is 
// copyrighted and licensed, not sold. You may use, copy, modify, and 
// distribute the Sample in any form without payment to IBM, for the purpose of 
// assisting you in the development of your applications.
// 
// The Sample code is provided to you on an "AS IS" basis, without warranty of 
// any kind. IBM HEREBY EXPRESSLY DISCLAIMS ALL WARRANTIES, EITHER EXPRESS OR 
// IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF 
// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. Some jurisdictions do 
// not allow for the exclusion or limitation of implied warranties, so the above 
// limitations or exclusions may not apply to you. IBM shall not be liable for 
// any damages you suffer as a result of using, copying, modifying or 
// distributing the Sample, even if IBM has been advised of the possibility of 
// such damages.
//***************************************************************************
//
// SOURCE FILE NAME: TbOnlineInx.java
//
// SAMPLE: How to create and reorg indexes on a table
//
// SQL STATEMENTS USED:
//         INCLUDE 
//         CREATE INDEX 
//         DROP INDEX
//         REORG
//         LOCK
//
// JAVA 2 CLASSES USED:
//         Statement
//         ResultSet
//         File
//         FileWriter
//         Process
//         BufferedReader
//         InputStreamReader
//
// Classes used from Util.java are:
//         Db
//         JdbcException
//
// Compile: the utility file and the source file with:
//		javac Util.java
//		javac <filename>.java
//
// Run: java <filename> [<db_name>] <username> <pwd>
//	or
//     java <filename> [<db_name>] <server_name> <port_num> <username> <pwd>
//
// OUTPUT FILE: TbOnlineInx.out (available in the online documentation)
// Output will vary depending on the JDBC driver connectivity used.
//***************************************************************************
//
// For more information on the sample programs, see the README file.
//
// For information on developing JDBC applications, see the Application
// Development Guide.
//
// For information on using SQL statements, see the SQL Reference.
//
// For the latest information on programming, compiling, and running DB2
// applications, visit the DB2 application development website at
//     http://www.software.ibm.com/data/db2/udb/ad
//**************************************************************************/

import java.sql.*;
import java.lang.*;
import java.io.*;

public class TbOnlineInx
{
  public static void main(String argv[]) 
  {
    try
    {
      Db db = new Db(argv);

      System.out.println();
      System.out.println(
        "THIS SAMPLE SHOWS HOW TO CREATE AND REORG ONLINE INDEXES\n" +
        "ON TABLES.");
            
      // connect to the 'sample' database
      db.connect();
      
      // create online index on a table 
      createIndex(db.con);
      
      // reorg online index on a table 
      reorgIndex(db.con);
      
      // drop online index created
      dropIndex(db.con);
      
      // disconnect from the 'sample' database
      db.disconnect();
    }
    catch (Exception e)
    {
      JdbcException jdbcExc = new JdbcException(e);
      jdbcExc.handle();
    }
  } // main

  // How to create an index on a table with different levels
  // of access to the table like read-write, read-only, no access 
  static void createIndex(Connection conn) throws Exception
  {
  
    System.out.print(
      "\n-----------------------------------------------------------" +
      "\nUSE THE SQL STATEMENT\n" +
      "  CREATE INDEX\n" +
      "TO CREATE AN INDEX\n");

    // create an online index with read-write access to the table 
    System.out.print(
      "\nTo create an index on a table allowing read-write access\n" +
      "to the table, use the following SQL command:\n\n" +
      "  CREATE INDEX index1 ON employee (lastname ASC)\n");
    
    Statement stmt = conn.createStatement();
    stmt.executeUpdate("CREATE INDEX index1 ON employee (lastname ASC)");
    conn.commit();
  
    dropIndex(conn);
  
    // create an index on a table while allowing only read access to it 
    System.out.println(
      "\nTo create an index on a table allowing only read access\n" +
      "to the table, use the following two SQL commands:\n\n" +
      "  LOCK TABLE employee IN SHARE MODE\n" +
      "  CREATE INDEX index1 ON employee (lastname ASC)");
  
    stmt.executeUpdate("LOCK TABLE employee IN SHARE MODE");
    conn.commit(); 

    stmt.executeUpdate("CREATE INDEX index1 ON employee (lastname ASC)");
    conn.commit();

    dropIndex(conn);
   
    // create an online index allowing no access to the table
    System.out.println(
      "\nTo create an index on a table allowing no access to the \n" +
      "table (only uncommitted readers allowed), use the \n" +
      "following two SQL statements:\n\n" +
      "  LOCK TABLE employee IN EXCLUSIVE MODE\n" +
      "  CREATE INDEX index1 ON employee (lastname ASC)");
    
    stmt.executeUpdate("LOCK TABLE employee IN EXCLUSIVE MODE");
    conn.commit();

    stmt.executeUpdate("CREATE INDEX index1 ON employee (lastname ASC)"); 
    conn.commit();  
 
    stmt.close();
  } // createIndex 
  
  // Create 3 CLP files for REORG command with write, read and no access,
  // respectively.
  static void createFiles(Connection conn) throws Exception 
  {
    // get fully qualified name of the table
    String tableName = "EMPLOYEE"; 
    String schemaName = getSchemaName(conn, tableName);    
    String fullTableName = schemaName + "." + tableName;      
       
    // reorg command has to be executed with three different options, namely,
    // 'with write access', 'with read access' and 'with no access'
    String[] fileNames = { "ReorgCmdAllowWrite.db2", 
                           "ReorgCmdAllowRead.db2", 
                           "ReorgCmdAllowNone.db2" };
     
    String[] options = { " WRITE ACCESS",
                         " READ ACCESS",
                         " NO ACCESS" };

    for (int i = 0; i < 3; i++)
    { 
      // create a CLP file with the REORG command and execute the file
      File outputFile = new File(fileNames[i]);
      FileWriter out = new FileWriter(outputFile);
    
      out.write("CONNECT TO SAMPLE;\n");
      out.write("REORG INDEXES ALL FOR TABLE " + fullTableName + 
                " ALLOW" + options[i] + ";\n");
      out.write("CONNECT RESET;");
      out.close();   
        
      // on exit, delete the temporary files created 
      outputFile.deleteOnExit();                
    }
  } //createFiles  
 
  // How to reorg an index on a table with different levels of 
  // access to the table like read-write, read-only, no access 
  static void reorgIndex(Connection conn) 
  {
    System.out.print(
      "\n-----------------------------------------------------------\n" +
      "\nUSE THE SQL STATEMENT:\n"+
      "  REORG\n" +
      "TO REORGANIZE A TABLE OR INDEX\n");

    
    String[] fileNames = { "ReorgCmdAllowWrite.db2", 
                           "ReorgCmdAllowRead.db2", 
                           "ReorgCmdAllowNone.db2" };
    
    String[] options = { " write access",
                         " read access",
                         " no access" };
    try
    {   
      // create 3 files with REORG commands
      createFiles(conn);
      
      for (int i = 0; i < 3; i++)
      { 
        System.out.println(
          "\nReorganize the indexes on a table allowing" + options[i] + 
          "\n-----------------------------------------------------------");
                 
        String s = null;  
        String execCmd = "db2 -tvf " + fileNames[i];
              
        // execute the command to run the CLP file
        Process p = Runtime.getRuntime().exec(execCmd);           
                                            
        // open streams for the process's input and error  
        BufferedReader stdInput = new BufferedReader(new 
                                        InputStreamReader(p.getInputStream()));
        BufferedReader stdError = new BufferedReader(new
                                        InputStreamReader(p.getErrorStream()));
    
        // read the output from the command and set the output variable with 
        // the value
        while ((s = stdInput.readLine()) != null)
        {
          System.out.println(s);
        }
      
        // read any errors from the attempted command and set the error  
        // variable with the value
        while ((s = stdError.readLine()) != null) 
        {
          System.out.println(s);
        }
     
        p.destroy();
              
      } // for            
    } // try
    
    catch (IOException e)
    {
      e.printStackTrace();
      System.exit(-1);             
    }  
     
    catch (Exception e)
    {
      JdbcException jdbcExc = new JdbcException(e);
      jdbcExc.handle();
    }       
         
  } // reorgIndex
  
  // How to drop the index on a table 
  static void dropIndex(Connection conn)
  {
    System.out.println(
      "\nUSE THE SQL STATEMENT\n" +
      "  DROP\n" +
      "TO DROP AN INDEX:");
    try
    {
      // drop the indexes 
      System.out.println(
        "  Execute the statement\n" +
        "    DROP INDEX index1\n" +
        "\n-----------------------------------------------------------");
      Statement stmt = conn.createStatement();
      stmt.executeUpdate("DROP INDEX index1");
      conn.commit();
      stmt.close();
    }
    catch (Exception e)
    {
      JdbcException jdbcExc = new JdbcException(e);
      jdbcExc.handle();
    }
  } // dropIndex 

  // function to get the schema name for a particular table
  static String getSchemaName(Connection conn, String tableName) throws Exception
  {
    Statement stmt = conn.createStatement();
    ResultSet rs = stmt.executeQuery(
                     "SELECT tabschema "+
                     "  FROM syscat.tables "+
                     "  WHERE tabname = '"+ tableName + "'");
                     
    boolean result = rs.next();
    String schemaName = rs.getString("tabschema");
    rs.close();
    stmt.close();
    
    // remove the trailing white space characters from schemaName before 
    // returning it to the calling function
    return schemaName.trim();
  } // getSchemaName  
} // TbOnlineInx    
